rdkPlugins/Storage/source/MountOwnerDetails.cpp

```diff
@@ -0,0 +1,241 @@
+ *
+ *  @param[out]  userId  ID corresponding to the configured user name
+ *  @param[out]  groupId  ID corresponding to the configured group name
+ *  @return true on success, false on failure
+ */
+bool MountOwnerDetails::getOwnerIds(uid_t& userId, gid_t& groupId) const
+{
+    struct passwd *pwd;
+    struct group *grp;
+    if (mMountOwnerProperties.user.empty())
+    {
+        AI_LOG_INFO("Using default user '%d'", mDefaultUserId);
+        userId = mDefaultUserId;
+    }
+    else
+    {
+        pwd = getpwnam(mMountOwnerProperties.user.c_str());
+        if (pwd == NULL)
+        {
+            AI_LOG_SYS_ERROR(errno, "User '%s' not found", mMountOwnerProperties.user.c_str());
+            return false;
+        }
+        else
+        {
+            userId = pwd->pw_uid;
+        }
+    }
+
+    if (mMountOwnerProperties.group.empty())
+    {
+        AI_LOG_INFO("Using default group '%d'", mDefaultGroupId);
+        groupId = mDefaultGroupId;
+    }
+    else
+    {
+        grp = getgrnam(mMountOwnerProperties.group.c_str());
+        if (grp == NULL)
+        {
+            AI_LOG_SYS_ERROR(errno, "Group '%s' not found", mMountOwnerProperties.group.c_str());
+            return false;
+        }
+        else
+        {
+            groupId = grp->gr_gid;
+        }
+    }
+
+    return true;
+}
+
+// -----------------------------------------------------------------------------
+/**
+ *  @brief Process individual mount owner and change ownership either singularly
+ * or recursively
+ *
+ *  @return true on success, false on failure
+ */
+bool MountOwnerDetails::processOwners() const
+{
+    AI_LOG_FN_ENTRY();
+
+    uid_t userId = 0;
+    gid_t groupId = 0;
+
+    if (getOwnerIds(userId, groupId))
+    {
+        if (mMountOwnerProperties.recursive)
+        {
+            if (!changeOwnerRecursive(mMountOwnerProperties.source, userId, groupId))
+            {
+                AI_LOG_ERROR("Failed to change owner recursively of '%s' to '%d:%d",
+                             mMountOwnerProperties.source.c_str(),
+                             userId,
+                             groupId);
+                return false;
+            }
+        }
+        else
+        {
+            changeOwner(mMountOwnerProperties.source, userId, groupId);
+        }
+    }
+
+    AI_LOG_FN_EXIT();
+    return true;
+}
+
+// -----------------------------------------------------------------------------
+/**
+ *  @brief Change ownership recursively from the given path
+ *
+ *  @param[out]  path    Path to recurse from
+ *  @param[out]  userId  ID corresponding to the configured user name
+ *  @param[out]  groupId ID corresponding to the configured group name
+ *  @return true on success, false on failure
+ */
+bool MountOwnerDetails::changeOwnerRecursive(const std::string& path, uid_t userId, gid_t groupId) const
+{
+    bool success = true;
+    DIR* dir = opendir(path.c_str());
+    if (!dir) {
+        return success;
```

Comment:
You are setting success initially to true, then if directory doesn't exist you are succeeding? Is this intentional? Shouldn't we return false as we didn't change the ownership (as it doesn't exist). If this is intentional please add this info into the `@return` field i.e. `true on success or non-existing directory, false on failure` But still I think that failing it would make more sense.

Reply:
It defaults to true for the purpose of the &= usage.
However, I missed that false should be returned if path is not a directory.
Fixed.